File: src/Async/SodaFlow.Core.Async.Tests/TestUtil.cs
using System;
using System.Collections.Concurrent;
using System.Diagnostics;
using System.Threading;
using System.Threading.Tasks;
using JetBrains.Annotations;

namespace SodaFlow.Async.Tests;

internal static class TestUtil
{
    /// <summary>Reads <paramref name="condition" /> until it is true, or fails the test at a
    /// timeout.</summary>
    public static void WaitUntil([InstantHandle] Func<bool> condition, int timeoutMs = 5000)
    {
        Stopwatch stopwatch = Stopwatch.StartNew();

        while (!condition())
        {
            if (stopwatch.ElapsedMilliseconds > timeoutMs)
            {
                throw new TimeoutException("Condition was not met within the timeout.");
            }

            Thread.Sleep(10);
        }
    }
}

/// <summary>
///     An async operation, with the input as its key, whose end a test controls with
///     <see cref="Release" /> and <see cref="Fail" />. Thus, a test does not race the true clock.
///     It also records the inputs that the pipeline called. Thus, a test can show that an operation
///     has the Running status, and not only an admission, before the release. That difference
///     shows the behavior of Queue against the behavior of Parallel.
/// </summary>
internal sealed class ControlledOperation<TInput, TResult>
    where TInput : notnull
{
    private readonly ConcurrentDictionary<TInput, TaskCompletionSource<TResult>> gates = new();
    private readonly ConcurrentDictionary<TInput, bool> startedInputs = new();

    public MapAsyncOperation<TInput, TResult> Operation => this.Run;

    public bool HasStarted(TInput input) => this.startedInputs.ContainsKey(input);

    public void Release(TInput input, TResult result) => this.GateFor(input).TrySetResult(result);

    public void Fail(TInput input, Exception error) => this.GateFor(input).TrySetException(error);

    private async Task<MapAsyncResult<TResult>> Run(
        TInput input,
        ResultFactory<TResult> resultFactory,
        CancellationToken token)
    {
        this.startedInputs[input] = true;

        TaskCompletionSource<TResult> tcs = this.GateFor(input);

        // ReSharper disable once UseAwaitUsing - CancellationTokenRegistration is only
        // IAsyncDisposable from .NET Core 3.0, and this compiles for net472 as well.
        using CancellationTokenRegistration registration =
            token.Register(() => tcs.TrySetCanceled(token));

        TResult result = await tcs.Task.ConfigureAwait(false);

        return resultFactory.FromValue(result);
    }

    private TaskCompletionSource<TResult> GateFor(TInput input) =>
        this.gates.GetOrAdd(key: input, valueFactory: static _ => new TaskCompletionSource<TResult>());
}

File: src/Async/FSharp/SodaFlow.FSharp.Async.Tests/TestUtil.fs
module SodaFlow.Async.Tests.TestUtil

open System
open System.Collections.Concurrent
open System.Diagnostics
open System.Threading
open System.Threading.Tasks
open SodaFlow.Async

/// Reads `condition` until it is true, or fails the test at a timeout.
let waitUntil (condition: unit -> bool) =
    let sw = Stopwatch.StartNew()

    while not (condition ()) do
        if sw.ElapsedMilliseconds > 5000L then
            raise (TimeoutException("Condition was not met within the timeout."))

        Thread.Sleep(10)

/// An async operation, with the input as its key, whose end a test controls with Release and
/// Fail. Thus, a test does not race the true clock. It also records the inputs that the pipeline
/// called. Thus, a test can show that an operation has the Running status, and not only an
/// admission, before the release.
type ControlledOperation<'TInput, 'TResult when 'TInput: equality>() =
    let gates = ConcurrentDictionary<'TInput, TaskCompletionSource<'TResult>>()
    let started = ConcurrentDictionary<'TInput, bool>()

    let gateFor input =
        gates.GetOrAdd(input, fun _ -> TaskCompletionSource<'TResult>())

    member _.HasStarted(input: 'TInput) = started.ContainsKey(input)

    member _.Release(input: 'TInput, result: 'TResult) =
        (gateFor input).TrySetResult(result) |> ignore

    member _.Fail(input: 'TInput, error: exn) =
        (gateFor input).TrySetException(error) |> ignore

    member _.Operation
        : 'TInput -> ResultFactory<'TResult> -> CancellationToken -> Task<MapAsyncResult<'TResult>> =
        fun input resultFactory token ->
            started[input] <- true
            let tcs = gateFor input
            token.Register(fun () -> tcs.TrySetCanceled(token) |> ignore) |> ignore
            tcs.Task.ContinueWith(fun (t: Task<'TResult>) -> resultFactory.FromValue(t.Result))
